src/MatrixClient.cc

```diff
@@ -1021,20 +1027,19 @@ MatrixClient::makeUploadRequest(const QString &filename)
         endpoint.setPath(mediaApiUrl_ + "/upload");
         endpoint.setQuery(query);
 
-        QFile file(filename);
-        if (!file.open(QIODevice::ReadWrite)) {
-                qDebug() << "Error while reading" << filename;
+        if (!iodev->open(QIODevice::ReadWrite)) {
+                qDebug() << "Error while reading buffer" << iodev.data();
                 return nullptr;
         }
 
         QMimeDatabase db;
-        QMimeType mime = db.mimeTypeForFile(filename, QMimeDatabase::MatchContent);
+        QMimeType mime = db.mimeTypeForData(iodev.data());
 
         QNetworkRequest request(QString(endpoint.toEncoded()));
-        request.setHeader(QNetworkRequest::ContentLengthHeader, file.size());
+        request.setHeader(QNetworkRequest::ContentLengthHeader, iodev->size());
         request.setHeader(QNetworkRequest::ContentTypeHeader, mime.name());
 
-        auto reply = post(request, file.readAll());
+        auto reply = post(request, iodev->readAll());
```

Comment:
It's inappropriate to read the `QIODevice` fully into memory before transmitting. This wastes time and memory. Instead, it should be streamed directly. Note that [`QNetworkManager::post`](https://doc.qt.io/qt-5/qnetworkaccessmanager.html#post) supports this directly.

Reply:
I was not aware of that particular overload. Fixed.

Reply:
I am having trouble with this one actually. It seems that `QNetworkAccessManager::post` when taking in a pointer of `QIODevice`, [will try to create another `QSharedPointer`](https://code.woboq.org/qt5/qtbase/src/corelib/io/qnoncontiguousbytedevice.cpp.html#_ZN31QNonContiguousByteDeviceFactory12createSharedEP9QIODevice). This causes problems, because in our case, the pointer to `QIODevice` is already a `QSharedPointer`. Somewhere this causes a segfault because of a double-free. 

I was able to reproduce the issue. The following code is based on the stack trace from gdb.

Here's the code:
```cpp
#include <QBuffer>
#include <QFile>
#include <QPointer>
#include <iostream>
#include <memory>

static void not_deleter(QObject *obj)
{
}

static void deleter(QObject *obj)
{
    delete obj;
}

auto foo(QIODevice *p)
{
    std::cerr << "Before cast\n";
    QBuffer *buf = qobject_cast<QBuffer*>(p);
    std::cerr << "After cast\n";

    std::cerr << "Before unique_ptr ctor\n";
    QSharedPointer<QIODevice>{p, &deleter};      // segfault
    QSharedPointer<QIODevice>{p, &not_deleter};  // OK
    std::cerr << "After unique_ptr ctor\n";
}

int main()
{
    std::cerr << "Begin main()\n";
    QSharedPointer<QFile> ptr{new QFile{}};

    std::cerr << "Before call to foo()\n";
    foo(ptr.data());
    std::cerr << "After call to foo()\n";

    std::cerr << "End main()\n";
    return 0;
}
```

Here's the stack trace:
```
Thread 1 "nheko" received signal SIGSEGV, Segmentation fault.
0x0000000e00000001 in ?? ()
(gdb) bt
#0  0x0000000e00000001 in  ()
#1  0x00007ffff5ea1b29 in QMetaObject::cast(QObject const*) const ()
    at /usr/lib/libQt5Core.so.5
#2  0x00007ffff5db9e7d in QNonContiguousByteDeviceFactory::createShared(QIODevice*) ()
    at /usr/lib/libQt5Core.so.5
#3  0x00007ffff6b39ecc in  () at /usr/lib/libQt5Network.so.5
#4  0x00007ffff6b40330 in  () at /usr/lib/libQt5Network.so.5
#5  0x00007ffff6b41c0c in  () at /usr/lib/libQt5Network.so.5
#6  0x00007ffff6b41cf5 in  () at /usr/lib/libQt5Network.so.5
#7  0x00007ffff6bf4fda in  () at /usr/lib/libQt5Network.so.5
#8  0x00007ffff5ec9452 in QObject::event(QEvent*) () at /usr/lib/libQt5Core.so.5
#9  0x00007ffff76d1e3c in QApplicationPrivate::notify_helper(QObject*, QEvent*) ()
    at /usr/lib/libQt5Widgets.so.5
#10 0x00007ffff76d9816 in QApplication::notify(QObject*, QEvent*) ()
    at /usr/lib/libQt5Widgets.so.5
#11 0x00007ffff5e981e0 in QCoreApplication::notifyInternal2(QObject*, QEvent*) ()
    at /usr/lib/libQt5Core.so.5
#12 0x00007ffff5e9ae46 in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () at /usr/lib/libQt5Core.so.5
#13 0x00007ffff5ef50a4 in  () at /usr/lib/libQt5Core.so.5
#14 0x00007ffff2670270 in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#15 0x00007ffff2671f69 in  () at /usr/lib/libglib-2.0.so.0
#16 0x00007ffff2671fae in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#17 0x00007ffff5ef4691 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Core.so.5
#18 0x00007fffe8f2c152 in  () at /usr/lib/libQt5XcbQpa.so.5
#19 0x00007ffff5e9682b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) ()
    at /usr/lib/libQt5Core.so.5
#20 0x00007ffff5e9fb18 in QCoreApplication::exec() () at /usr/lib/libQt5Core.so.5
#21 0x0000555555a21ee7 in main(int, char**) (argc=1, argv=0x7fffffffe028)
    at /home/chris/dev/nheko/src/main.cc:166
```
See the link above for the source code of `QNonContiguousByteDeviceFactory::createShared`.

Reply:
Referring to [the documentation of `QSharedPointer::create`](https://doc.qt.io/qt-5/qsharedpointer.html#create), you can see that it does not create a new `QSharePointer` owning exactly the same object, but rather creates a new, unrelated `QSharedPointer` with its own object. This makes sense, because (although the weird syntax highlighting makes it a bit hard to see) the value being passed in is a pointer to the pointer, not the pointer itself.

[The documentation for `QNetworkAccessManager::post`](https://doc.qt.io/qt-5/qnetworkaccessmanager.html#post) specifies that you must ensure the `QIODevice` remains valid until the returned reply emits its `finished` signal. If you're seeing memory errors, this may be the cause. I believe a convenient way to ensure the `QIODevice` lives that long would be attaching a closure that captures the `QSharedPointer` that owns it to the signal.

Reply:
Ah, thanks. Yes that fixed the issue.